File: 1161.maximum-level-sum-of-a-binary-tree.py
#
# @lc app=leetcode id=1161 lang=python3
#
# [1161] Maximum Level Sum of a Binary Tree
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        # BFS
        if not root:
            return 0
        ans = (1, root.val)  # level, bestSum
        queue = [root]
        level = 1
        while queue:
            _nextQ = []
            _sum = 0
            for node in queue:
                _sum += node.val
                if node.left:
                    _nextQ.append(node.left)
                if node.right:
                    _nextQ.append(node.right)
            if _sum > ans[1]:
                ans = (level, _sum)
            queue = _nextQ
            level += 1
        return ans[0]


# @lc code=end

File: leetcode-nvim/1161.maximum-level-sum-of-a-binary-tree.go
package main

import "fmt"

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type NodeLevelPair struct {
	Node  *TreeNode
	Level int
}

func maxLevelSum(root *TreeNode) int {
	if root == nil {
		return 0
	}
	ans := []int{1, root.Val} // level, val pair
	queue := []NodeLevelPair{}
	queue = append(queue, NodeLevelPair{Node: root, Level: 1})
	level2Sum := map[int]int{}
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		level2Sum[curr.Level] += curr.Node.Val
		if curr.Node.Left != nil {
			queue = append(queue, NodeLevelPair{Node: curr.Node.Left, Level: curr.Level + 1})
		}
		if curr.Node.Right != nil {
			queue = append(queue, NodeLevelPair{Node: curr.Node.Right, Level: curr.Level + 1})
		}
		if len(queue) == 0 || (len(queue) > 0 && queue[0].Level != curr.Level) {
			if level2Sum[curr.Level] > ans[1] {
				fmt.Println(level2Sum[curr.Level], curr.Level)
				ans[0] = curr.Level
				ans[1] = level2Sum[curr.Level]
			}
		}
	}
	return ans[0]
}

// @leet end

